design/one-pager-trait-workload-interaction-mechanism.md

```diff
@@ -0,0 +1,198 @@
+an **optional** field called `childResources` to the schema of the workloadDefinition. We encourage
+workload owners to fill in this field when they register their controllers to the OAM system. 
+This is the way for them to declare the types of the Kubernetes resources their workload
+controller actually generates. In our example, the workload definition can claim to generate
+deployment and service child resources. 
+    ```yaml
+    apiVersion: core.oam.dev/v1alpha2
+    kind: WorkloadDefinition
+    metadata:
+       name: mydb.standard.oam.dev
+    spec:
+      definitionRef:
+         name: mydb.standard.oam.dev
+      childResources:
+         - groupVersion: apps/v1
+           kind: deployment
+         - groupVersion: core/v1
+           kind: service      
+    ```
+
+2. ApplicationConfig controller add annotations containing workload child resource information to
+the emitted `workload` CR. For each `workload` CR, the applicationConfig controller can fetch the
+corresponding `workloadDefinition` CR following an 
+[OAM convention](https://github.com/oam-dev/spec/blob/master/3.workload.md#definitionref). The
+conventions says that the name of the `workloadDefinition` CR is the name of the `workload` CRD
+it refers to. For example, the name of the `workloadDefinition` CR that refers to a
+`containerizedworkloads.core.oam.dev` CRD is exactly `containerizedworkloads.core.oam.dev` as
+well. In our example, the applicationConfig controller will emit the following `workload` CR with all
+ its child resources GVK in its metadata.
+    ```yaml
+    apiVersion: standard.oam.dev/v1alpha2
+    kind: Mydb
+    metadata:
+      name: mydb-example
+      annotations:
+        core.oam.dev/childResources: "{ apps/v1.deployment, core/v1.service}"
+    ```
+
+3. ApplicationConfig controller add annotations containing the workload information to the emitted
+`trait` CR. The application controller actually generates the workload CR so it has all the
+information. In our example, the applicationConfig controller will emit the following `trait` CR with
+the GVK and name of the workload the `ManualScalerTrait` is applied to.
+    ```yaml
+    apiVersion: core.oam.dev/v1alpha2
+    kind: ManualScalerTrait
+    metadata:
+      annotations:
+        core.oam.dev/workloadGroupVersion: standard.oam.dev/v1alpha2
+        core.oam.dev/workloadKind: Mydb
+        core.oam.dev/workloadName: mydb-example
+    ```
+4. OAM runtime will provide a helper library. The library follows the following logic to help a
+ trait developer locate the resources for the trait to modify
+    1. Get the corresponding `workload` instance from the Kubernetes cluster with the information
+     contained in the annotation field inserted by the application controller in the `trait` CR.
+    2. Fetch all the `childResources` values inserted by the application controller in the
+     annotation field in the `workload` instance. 
```

Comment:
Would it be feasible to lookup the `WorkloadDefinition` that corresponds to the workload and read its `spec.childResources`, rather than requiring an annotation on the workload CR itself? This would avoid us having to repeat that information many times, and also avoid us storing it as unvalidated, opaque structured data (presumably JSON) inside an annotation.

Reply:
The annotation on the workload CR itself is for a `trait`. It seems that every trait needs to repeat what an appConfig controller does without it. We are hoping to save this part of the logic for the OAM core. Does it make sense?

Reply:
I agree it would be better to avoid having this logic distributed across all of the traits. I like the idea of having a library in `oam-kubernetes-runtime` that can help with this. My question is whether that library could grab this metadata from the `WorkloadDefinition` instead of an annotation.

I imagine this would look something like:

4. OAM runtime will provide a helper library. The library follows the following logic to help a trait developer locate the resources for the trait to modify.
    1. Read the `spec.workloadRef` of the trait to determine what kind of workload it applies to and fetch the `WorkloadDefinition` that defines the kind of workload.
    1. Read the `spec.childWorkloadKinds` of the `WorkloadDefinition`
    1. List each child resource by its GVK and filter by controller reference.

Does that make sense?

Reply:
I think what @negz  says is fine, but the workloadRef should be optional.

Reply:
yeah, that works. The only concern is that the workload to workloadDefinition part can be confusing so we wanted to hide that part from the user.